File: lib/mercuryapi-1.23.0.20/c/src/samples/readerInfo.c
/**
 * Sample program that display reader parameters
 * @file readerinfo.c
 */

#include <tm_reader.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>

/* Enable this to use transportListener */
#ifndef USE_TRANSPORT_LISTENER
#define USE_TRANSPORT_LISTENER 0
#endif

void errx(int exitval, const char *fmt, ...)
{
  va_list ap;

  va_start(ap, fmt);
  vfprintf(stderr, fmt, ap);

  exit(exitval);
}

void checkerr(TMR_Reader* rp, TMR_Status ret, int exitval, const char *msg)
{
  if (TMR_SUCCESS != ret)
  {
    errx(exitval, "Error %s: %s\n", msg, TMR_strerr(rp, ret));
  }
}

void serialPrinter(bool tx, uint32_t dataLen, const uint8_t data[],
                   uint32_t timeout, void *cookie)
{
  FILE *out = cookie;
  uint32_t i;

  fprintf(out, "%s", tx ? "Sending: " : "Received:");
  for (i = 0; i < dataLen; i++)
  {
    if (i > 0 && (i & 15) == 0)
    {
      fprintf(out, "\n         ");
    }
    fprintf(out, " %02x", data[i]);
  }
  fprintf(out, "\n");
}

void stringPrinter(bool tx,uint32_t dataLen, const uint8_t data[],uint32_t timeout, void *cookie)
{
  FILE *out = cookie;

  fprintf(out, "%s", tx ? "Sending: " : "Received:");
  fprintf(out, "%s\n", data);
}

int main(int argc, char *argv[])
{
  TMR_Reader r, *rp;
  TMR_Status ret;
  TMR_Region region;
#if USE_TRANSPORT_LISTENER
  TMR_TransportListenerBlock tb;
#endif
 
  if (argc < 2)
  {
    errx(1, "Please provide reader URL, such as:\n"
           "tmr:///com4\n"
           "tmr://my-reader.example.com\n");
  }
  
  rp = &r;
  ret = TMR_create(rp, argv[1]);
  checkerr(rp, ret, 1, "creating reader");

#if USE_TRANSPORT_LISTENER

  if (TMR_READER_TYPE_SERIAL == rp->readerType)
  {
    tb.listener = serialPrinter;
  }
  else
  {
    tb.listener = stringPrinter;
  }
  tb.cookie = stdout;

  TMR_addTransportListener(rp, &tb);
#endif

  ret = TMR_connect(rp);
  checkerr(rp, ret, 1, "connecting reader");

  region = TMR_REGION_NONE;
  ret = TMR_paramGet(rp, TMR_PARAM_REGION_ID, &region);
  checkerr(rp, ret, 1, "getting region");

  if (TMR_REGION_NONE == region)
  {
    TMR_RegionList regions;
    TMR_Region _regionStore[32];
    regions.list = _regionStore;
    regions.max = sizeof(_regionStore)/sizeof(_regionStore[0]);
    regions.len = 0;

    ret = TMR_paramGet(rp, TMR_PARAM_REGION_SUPPORTEDREGIONS, &regions);
    checkerr(rp, ret, __LINE__, "getting supported regions");

    if (regions.len < 1)
    {
      checkerr(rp, TMR_ERROR_INVALID_REGION, __LINE__, "Reader doesn't supportany regions");
    }
    region = regions.list[0];
    ret = TMR_paramSet(rp, TMR_PARAM_REGION_ID, &region);
    checkerr(rp, ret, 1, "setting region");  
  }

  {
    uint16_t productGroupID, productID;
    TMR_String str;
    char string[50];
    str.value = string;
    str.max = 50;

    ret = TMR_paramGet(rp, TMR_PARAM_VERSION_HARDWARE, &str);
    if (TMR_SUCCESS == ret)
    {
      printf("/reader/version/hardware: %s\n", str.value);
    }
    else
    {
      if (TMR_ERROR_NOT_FOUND == ret)
      {
        printf("/reader/version/hardware not supported\n");
      }
      else
      {
        printf("Error %s: %s\n", "getting version hardware",
            TMR_strerr(rp, ret));
      }
    }
    /**
     * for failure case API is modifying the str.value to some constant string,
     * to fix that, restoring the str.value variable
     **/ 
    str.value=string;

    ret = TMR_paramGet(rp, TMR_PARAM_VERSION_SERIAL, &str);
    if (TMR_SUCCESS == ret)
    {
      printf("/reader/version/serial: %s\n", str.value);
    }
    else
    {
      if (TMR_ERROR_NOT_FOUND == ret)
      {
        printf("/reader/version/serial not supported\n");
      }
      else
      {
        printf("Error %s: %s\n", "getting version serial",
            TMR_strerr(rp, ret));
      }
    }
    /**
     * for failure case API is modifying the str.value to some constant string,
     * to fix that, restoring the str.value variable
     **/ 
     str.value=string;

    ret = TMR_paramGet(rp, TMR_PARAM_VERSION_MODEL, &str);
    if (TMR_SUCCESS == ret)
    {
      printf("/reader/version/model:  %s\n", str.value);
    }
    else
    {
      if (TMR_ERROR_NOT_FOUND == ret)
      {
        printf("/reader/version/model not supported\n");
      }
      else
      {
        printf("Error %s: %s\n", "getting version model",
            TMR_strerr(rp, ret));
      }
    }
    /**
     * for failure case API is modifying the str.value to some constant string,
     * to fix that, restoring the str.value variable
     **/ 
     str.value=string;

    ret = TMR_paramGet(rp, TMR_PARAM_VERSION_SOFTWARE, &str);
    if (TMR_SUCCESS == ret)
    {
      printf("/reader/version/software: %s\n", str.value);
    }
    else
    {
      if (TMR_ERROR_NOT_FOUND == ret)
      {
        printf("/reader/version/software not supported\n");
      }
      else
      {
        printf("Error %s: %s\n", "getting version software",
            TMR_strerr(rp, ret));
      }
    }
    /**
     * for failure case API is modifying the str.value to some constant string,
     * to fix that, restoring the str.value variable
     **/ 
     str.value=string;

    ret = TMR_paramGet(rp, TMR_PARAM_URI, &str);
    if (TMR_SUCCESS == ret)
    {
      printf("/reader/uri:  %s\n",str.value);
    }
    else
    {
      if (TMR_ERROR_NOT_FOUND == ret)
      {
        printf("/reader/uri:  Unsupported\n");
      }
      else
      {
        printf("Error %s: %s\n", "getting reader URI",
            TMR_strerr(rp, ret));
      }
    }
    /**
     * for failure case API is modifying the str.value to some constant string,
     * to fix that, restoring the str.value variable
     **/ 
     str.value=string;

    ret = TMR_paramGet(rp, TMR_PARAM_PRODUCT_ID, &productID);
    if (TMR_SUCCESS == ret)
    {
      printf("/reader/version/productID: %d\n", productID);
    }
    else
    {
      if (TMR_ERROR_NOT_FOUND == ret)
      {
        printf("/reader/version/productID not supported\n");
      }
      else
      {
        printf("Error %s: %s\n", "getting product id",
            TMR_strerr(rp, ret));
      }
    }

    ret = TMR_paramGet(rp, TMR_PARAM_PRODUCT_GROUP_ID, &productGroupID);
    if (TMR_SUCCESS == ret)
    {
      printf("/reader/version/productGroupID: %d\n", productGroupID);
    }
    else
    {
      if (TMR_ERROR_NOT_FOUND == ret)
      {
        printf("/reader/version/productGroupID not supported\n");
      }
      else
      {
        printf("Error %s: %s\n", "getting product group id",
            TMR_strerr(rp, ret));
      }
    }

    ret = TMR_paramGet(rp, TMR_PARAM_PRODUCT_GROUP, &str);
    if (TMR_SUCCESS == ret)
    {
      printf("/reader/version/productGroup: %s\n", str.value);
    }
    else
    {
      if (TMR_ERROR_NOT_FOUND == ret)
      {
        printf("/reader/version/productGroup not supported\n");
      }
      else
      {
        printf("Error %s: %s\n", "getting product group",
            TMR_strerr(rp, ret));
      }
    }
    /**
     * for failure case API is modifying the str.value to some constant string,
     * to fix that, restoring the str.value variable
     **/ 
     str.value=string;

    ret = TMR_paramGet(rp, TMR_PARAM_READER_DESCRIPTION, &str);
    if (TMR_SUCCESS == ret)
    {
      printf("/reader/description:  %s\n", str.value);
    }
    else
    {
      if (TMR_ERROR_NOT_FOUND == ret)
      {
        printf("/reader/description not supported\n");
      }
      else
      {
        printf("Error %s: %s\n", "getting reader description",
            TMR_strerr(rp, ret));
      }
    }
  }

  TMR_destroy(rp);
  return 0;
}

File: lib/mercuryapi-1.23.0.20/java/samples_nb/src/samples/ReaderInfo.java
/*
 * Sample program that display reader parameters
 */
package samples;

import com.thingmagic.Reader;
import com.thingmagic.TMConstants;
import com.thingmagic.TransportListener;

/**
 *
 * @author qvantel
 */
public class ReaderInfo
{
    static void usage()
    {
        System.out.printf("Please provide reader URL, such as:\n"
                + "  (URI: 'tmr:///COM1' or 'tmr://astra-2100d3/' "
                + "or 'tmr:///dev/ttyS0')\n\n");
        System.exit(1);
    }

    public static void setTrace(Reader r, String args[])
    {
        if (args[0].toLowerCase().equals("on"))
        {
            r.addTransportListener(r.simpleTransportListener);
        }
    }

    static class StringPrinter implements TransportListener
    {
        public void message(boolean tx, byte[] data, int timeout)
        {
            System.out.println((tx ? "Sending:\n" : "Receiving:\n")
                    + new String(data));
        }
    }

    static class SerialPrinter implements TransportListener
    {
        public void message(boolean tx, byte[] data, int timeout)
        {
            System.out.print(tx ? "Sending: " : "Received:");
            for (int i = 0; i < data.length; i++)
            {
                if (i > 0 && (i & 15) == 0) {
                    System.out.printf("\n         ");
                }
                System.out.printf(" %02x", data[i]);
            }
            System.out.printf("\n");
        }
    }

    public static void main(String argv[])
    {
        Reader r = null;
        int nextarg = 0;
        boolean trace = false;

        if (argv.length < 1)
        {
            usage();
        }

        if (argv[nextarg].equals("-v"))
        {
            trace = true;
            nextarg++;
        }

        try
        {
            r = Reader.create(argv[nextarg]);
            if (trace)
            {
                setTrace(r, new String[]{"on"});
            }
            r.connect();
            if (Reader.Region.UNSPEC == (Reader.Region) r.paramGet(TMConstants.TMR_PARAM_REGION_ID))
            {
                Reader.Region[] supportedRegions = (Reader.Region[]) r.paramGet(TMConstants.TMR_PARAM_REGION_SUPPORTEDREGIONS);
                if (supportedRegions.length < 1) {
                    throw new Exception("Reader doesn't support any regions");
                } else {
                    r.paramSet(TMConstants.TMR_PARAM_REGION_ID, supportedRegions[0]);
                }
            }

            try
            {
                String version_hardware = (String) r.paramGet(TMConstants.TMR_PARAM_VERSION_HARDWARE);
                System.out.println("Hardware Version :" + version_hardware);
            } catch (Exception ex) {
                System.out.println("Hardware Version :" + ex.getMessage());
            }

            try
            {
                String version_serial = (String) r.paramGet(TMConstants.TMR_PARAM_VERSION_SERIAL);
                System.out.println("Serial Version :" + version_serial);
            } catch (Exception ex) {
                System.out.println("Serial Version :" + ex.getMessage());
            }
            try
            {
                String version_model = (String) r.paramGet(TMConstants.TMR_PARAM_VERSION_MODEL);
                System.out.println("Model Version  :" + version_model);
            } catch (Exception ex) {
                System.out.println("Model Version :" + ex.getMessage());
            }

            try
            {
                String version_software = (String) r.paramGet(TMConstants.TMR_PARAM_VERSION_SOFTWARE);
                System.out.println("Software Version :" + version_software);
            } catch (Exception ex) {
                System.out.println("Software Version :" + ex.getMessage());
            }

            try
            {
                String reader_uri = (String) r.paramGet(TMConstants.TMR_PARAM_READER_URI);
                System.out.println("Reader Uri :" + reader_uri);
            } catch (Exception ex) {
                System.out.println("Reader Uri :" + ex.getMessage());
            }

            try
            {
                String reader_productgroupid = (String) r.paramGet(TMConstants.TMR_PARAM_READER_PRODUCTGROUPID);
                System.out.println("Reader Product GroupId :" + reader_productgroupid);
            } catch (Exception ex) {
                System.out.println("Reader Product GroupId :" + ex.getMessage());
            }

            try
            {
                String reader_productgroup = (String) r.paramGet(TMConstants.TMR_PARAM_READER_PRODUCTGROUP);
                System.out.println("Reader Product Group :" + reader_productgroup);
            } catch (Exception ex) {
                System.out.println("Reader Product Group :" + ex.getMessage());
            }

            try
            {
                String reader_productid = (String) r.paramGet(TMConstants.TMR_PARAM_READER_PRODUCTID);
                System.out.println("Reader Product Id :" + reader_productid);
            } catch (Exception ex) {
                System.out.println("Reader Product Id :" + ex.getMessage());
            }

            try
            {
                String reader_description = (String) r.paramGet(TMConstants.TMR_PARAM_READER_DESCRIPTION);
                System.out.println("Reader Description :" + reader_description);
            } catch (Exception ex) {
                System.out.println("Reader Description :" + ex.getMessage());
            }
            // Shut down reader
            r.destroy();
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }
}
